armsrc/pcf7931.c

```diff
@@ -141,245 +140,364 @@ int DemodPCF7931(uint8_t **outBlocks) {
 	return num_blocks;
 }
 
-int IsBlock0PCF7931(uint8_t *Block) {
-	// Assume RFU means 0 :)
-	if((memcmp(Block, "\x00\x00\x00\x00\x00\x00\x00\x01", 8) == 0) && memcmp(Block+9, "\x00\x00\x00\x00\x00\x00\x00", 7) == 0) // PAC enabled
-		return 1;
-	if((memcmp(Block+9, "\x00\x00\x00\x00\x00\x00\x00", 7) == 0) && Block[7] == 0) // PAC disabled, can it *really* happen ?
-		return 1;
+uint8_t IsBlock0PCF7931(uint8_t *block)
+{
+	// assuming all RFU bits are set to 0
+	// if PAC is enabled password is set to 0
+	if (block[7] == 0x01)
+	{
+		if (!memcmp(block, "\x00\x00\x00\x00\x00\x00\x00", 7) && !memcmp(block+9, "\x00\x00\x00\x00\x00\x00\x00", 7))
+			return 1;
+	}
+	else if (block[7] == 0x00)
+	{
+		if (!memcmp(block+9, "\x00\x00\x00\x00\x00\x00\x00", 7))
+			return 1;
+	}
 	return 0;
 }
 
-int IsBlock1PCF7931(uint8_t *Block) {
-	// Assume RFU means 0 :)
-	if(Block[10] == 0 && Block[11] == 0 && Block[12] == 0 && Block[13] == 0)
-		if((Block[14] & 0x7f) <= 9 && Block[15] <= 9)
+uint8_t IsBlock1PCF7931(uint8_t *block) {
+	// assuming all RFU bits are set to 0
+	if (block[10] == 0 && block[11] == 0 && block[12] == 0 && block[13] == 0)
+		if((block[14] & 0x7f) <= 9 && block[15] <= 9)
 			return 1;
 
 	return 0;
 }
 
-void ReadPCF7931() {
-	uint8_t Blocks[8][17];
-	uint8_t tmpBlocks[4][16];
-	int i, j, ind, ind2, n;
-	int num_blocks = 0;
-	int max_blocks = 8;
-	int ident = 0;
-	int error = 0;
-	int tries = 0;
-
-	memset(Blocks, 0, 8*17*sizeof(uint8_t));
-
-	do {
-		memset(tmpBlocks, 0, 4*16*sizeof(uint8_t));
-		n = DemodPCF7931((uint8_t**)tmpBlocks);
+static inline void print_block(uint8_t *data)
+{
+	Dbprintf("%02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x",
+		data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7],
+		data[8], data[9], data[10], data[11], data[12], data[13], data[14], data[15]);
+}
+
+void ReadPCF7931()
+{
+	int found_blocks = 0; // successfully read blocks
+	int max_blocks = 8;	// readable blocks	
+	uint8_t memory_blocks[8][17]; // PCF content
+	
+	uint8_t single_blocks[8][17]; // PFC blocks with unknown position
+	int single_blocks_cnt = 0;
+
+	size_t n = 0; // transmitted blocks	
+	uint8_t tmp_blocks[4][16]; // temporary read buffer
+	
+	uint8_t found_0_1 = 0; // flag: blocks 0 and 1 were found
+	int errors = 0; // error counter
+	int tries = 0; // tries counter
+	
+	memset(memory_blocks, 0, 8*17*sizeof(uint8_t));
+	memset(single_blocks, 0, 8*17*sizeof(uint8_t));
+	
+	int i = 0, j = 0;
+
+	do
+	{
+		i = 0;
+		
+		memset(tmp_blocks, 0, 4*16*sizeof(uint8_t));
+		n = DemodPCF7931((uint8_t**)tmp_blocks);
 		if(!n)
-			error++;
-		if(error==10 && num_blocks == 0) {
+			errors++;
+		
+		// exit if no block is received 
+		if (errors == 10 && found_blocks == 0)
+		{
 			Dbprintf("Error, no tag or bad tag");
 			return;
 		}
-		else if (tries==20 || error==10) {
+		// exit if too many errors during reading
+		else if (tries > 50 && (float)(errors/tries) > .5f)
```

Comment:
`(float)(errors/tries) > .5f`: not that good. This will force the linker to add floating point library functions to the ARM code and unnecessarily increase the code size.  Please stick to integer, e.g. `errors > 2 * tries`